<a href="https://colab.research.google.com/github/thearyankarthik/thearyankarthik/blob/main/M1_Assignment_Stock_Price_Prediction_with_LLM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Module 1 No Code/Low Code Assignment: Stock Price Prediction with LLM (Using Gemini in Colab)

This assignment explores how large language models (LLMs) can perform complex tasks that previously required coding. Instead of manually writing code, we will interact with Google's Gemini (Click on the Gemini icon on the top right corner) in Google Colab to:

- Load and analyze stock price data.
- Engineer relevant features for prediction.
- Build and evaluate models such as regression and random forest.
- Discuss potential trading strategies based on predictions.

**Note**: While this notebook contains example code, students are encouraged to use LLMs to generate code snippets and explanations, managing interactions within the LLM's context window.

## Task Breakdown

### Step 1: Data Exploration
- **Prompt the LLM**: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance.
- **Key Considerations**: Ensure the model retains context while summarizing key statistics.

### Step 2: Feature Engineering
- **Prompt the LLM**: Generate meaningful features such as moving averages, technical indicators, and volatility measures.
- **Managing Context**: If the LLM forgets prior outputs, remind it by reintroducing key statistics.

### Step 3: Model Building
- **Prompt the LLM**: Train a model (regression for price prediction, random forest for trend classification).
- **Split Ratio Consideration**: Explain the importance of train-test split (e.g., 70/30) and verify the LLM respects it.

### Step 4: Model Evaluation & Prediction
- **Prompt the LLM**: Use the trained model to predict stock prices for 2023-2024.
- **Analyze Performance**: Interpret accuracy, R² scores, and classification reports.

### Step 5: Strategy Discussion
- **Prompt the LLM**: Based on model performance, discuss possible trading strategies.
- **Critical Thinking**: Evaluate whether the model is truly useful for financial decision-making.

---

Throughout the assignment, keep in mind that LLMs have a limited context window. If necessary, break down tasks into smaller chunks and reintroduce critical information.



## Using Gemini in Colab

To use Gemini directly in your browser within Google Colab:

1. **Create a new code cell** by clicking the `+ Code` button.
2. In the newly created cell, you will see a placeholder text: **"Start coding or generate with AI"**.
3. Click on **"Generate"** and enter your prompt, such as:
   > "Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics."
4. Gemini will generate code based on your prompt. You can then review, modify, and run it within the notebook.
5. If the model loses context, you may need to remind it of previous responses or break down tasks into smaller steps.

This method allows you to interactively generate and refine code without manually writing every line.


In [1]:
# prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics.

!pip install yfinance

import yfinance as yf
import pandas as pd

# Download historical data for AAPL
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Display summary statistics
print(data.describe())


/tmp/ipython-input-1-1433581267.py:9: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.633101  129.181519  125.971433  127.555151  1.120920e+08
std      30.421462   30.698899   30.139549   30.449562  5.602586e+07
min      54.378574   55.379527   51.528408   55.277736  3.519590e+07
25%     112.709509  114.204937  111.251838  112.889668  7.636470e+07
50%     133.010788  134.143870  131.050642  132.832299  9.493580e+07
75%     148.221519  149.388580  146.276881  147.588789  1.296327e+08
max     178.645615  179.558473  175.809076  179.254206  4.265100e+08


In [2]:
# prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics.

!pip install yfinance

import yfinance as yf
import pandas as pd

# Download historical data for AAPL from 2020-01-01 to 2023-01-01
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Summarize key statistics
print(data.describe())


/tmp/ipython-input-2-730687086.py:9: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.633101  129.181519  125.971433  127.555151  1.120920e+08
std      30.421462   30.698899   30.139549   30.449562  5.602586e+07
min      54.378574   55.379527   51.528408   55.277736  3.519590e+07
25%     112.709509  114.204937  111.251838  112.889668  7.636470e+07
50%     133.010788  134.143870  131.050642  132.832299  9.493580e+07
75%     148.221519  149.388580  146.276881  147.588789  1.296327e+08
max     178.645615  179.558473  175.809076  179.254206  4.265100e+08


In [3]:
# Prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01
# using Yahoo Finance and summarize its key statistics.

import yfinance as yf

data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Summarize key statistics
print(data.describe())



/tmp/ipython-input-3-3699561310.py:6: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.633101  129.181519  125.971433  127.555151  1.120920e+08
std      30.421462   30.698899   30.139549   30.449562  5.602586e+07
min      54.378574   55.379527   51.528408   55.277736  3.519590e+07
25%     112.709509  114.204937  111.251838  112.889668  7.636470e+07
50%     133.010788  134.143870  131.050642  132.832299  9.493580e+07
75%     148.221519  149.388580  146.276881  147.588789  1.296327e+08
max     178.645615  179.558473  175.809076  179.254206  4.265100e+08


## Now try doing step 2 - step 5 by yourself

The sample prompts and LLM responses are listed below for your reference.

In [4]:
import yfinance as yf
import pandas as pd

# Download historical data for AAPL from 2020-01-01 to 2023-01-01
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Summarize key statistics
print(data.describe())

/tmp/ipython-input-4-1219248934.py:5: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.633101  129.181519  125.971433  127.555151  1.120920e+08
std      30.421462   30.698899   30.139549   30.449562  5.602586e+07
min      54.378574   55.379527   51.528408   55.277736  3.519590e+07
25%     112.709509  114.204937  111.251838  112.889668  7.636470e+07
50%     133.010788  134.143870  131.050642  132.832299  9.493580e+07
75%     148.221519  149.388580  146.276881  147.588789  1.296327e+08
max     178.645615  179.558473  175.809076  179.254206  4.265100e+08


# Sample Prompts and LLM Responses
---

## Step 1: Data Exploration

### Prompt:
> Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 from Yahoo Finance, keep it in memory and summarize its key statistics.

### LLM Response:
- **Highest Closing Price:** \$182.01 on January 3, 2022
- **Lowest Closing Price:** \$56.01 on March 23, 2020
- **Average Closing Price:** \$126.69
- **Total Trading Days:** 756

---

## Step 2: Feature Engineering

### Prompt:
> Suggest some features that might help improve a stock price prediction model.

### LLM Response:
1. **Historical Prices**: Include past closing, opening, high, and low prices.
2. **Volume**: Daily trading volumes.
3. **Moving Averages**: Simple and exponential moving averages (e.g., 5-day, 10-day, 50-day).
4. **Technical Indicators**: RSI, MACD, Bollinger Bands, etc.
5. **Volatility**: Historical volatility, Average True Range (ATR).
6. **Sentiment Analysis**: News articles, social media sentiment.
7. **Economic Indicators**: Interest rates, inflation rates, GDP growth.
8. **Time-based Features**: Day of the week, month, year.
9. **Sector Performance**: Comparison with industry and sector indices.
10. **External Events**: Major geopolitical events, earnings reports.

---

## Step 3.1: Model Building - Linear Regression

### Prompt:
> Adopt the first four kinds of features and train a regression model to predict tomorrow's stock price. Split your training and testing data by a 70/30 ratio and report your performance metric.

### LLM Response:
- **Mean Squared Error (MSE):** 859.04
- **R-squared (R²) Score:** -0.009

These metrics indicate that the model is not performing well, as an R² score close to zero or negative suggests that the model does not explain the variability in the target variable better than a simple mean of the target variable. This outcome suggests that the chosen features and model might need adjustments or improvements to enhance prediction accuracy. ​​

---
(It might be too hard to predict the exact price but how about ups and downs?)
# Step 3.2: Model Building - Random Forest Classification

### Prompt:
> Create a binary target variable based on historical prices (1 if the close price increase tomorrow, 0 other), and train a Random Forest model using the same train/test split ratio. Report feature importance and performance metrics.

### LLM Response:
- **Accuracy:** 71.82%
- **Confusion Matrix:** [[84, 26],
 [36, 74]]
- **Classification Report:**

              precision    recall  f1-score   support

         0       0.70      0.76      0.73       110
         1       0.74      0.67      0.70       110
        accuracy                     0.72       220
        macro avg    0.72      0.72      0.72       220
        weighted avg  0.72      0.72      0.72       220



- **Feature Importance:** The most important feature for predicting whether the price will increase or decrease the next day is the closing price, followed by the 5-day moving average (MA_5). The model achieved an accuracy of approximately 71.82%, indicating a reasonable performance for this classification task.
         Feature	Importance
         Open	   0.0900
         High	   0.0883
         Low	   0.0980
         Close	   0.3087
         Volume	    0.0982
         MA_5	   0.1242
         MA_10	   0.0974
         MA_50	   0.0951

---

## Step 4: Model Evaluation & Prediction
### Prompt:
> Use this model to predict stock price increase/decrease for period starting from 2023-01-01 to 2024-01-01. Report your performance metrics.

### LLM Response:
The Random Forest classifier's performance on the new test data from 2023-01-01 to 2024-01-01 is as follows:
- **Accuracy:** 60.85%
- **Confusion Matrix:** [[105,   0],
 [ 83,  24]]
- **Classification Report:**

            precision    recall  f1-score   support

       0       0.56      1.00      0.72       105
       1       1.00      0.22      0.37       107

        accuracy                   0.61       212
        macro avg  0.78     0.61       0.54       212
        weighted avg 0.78    0.61        0.54       212

---
## Step 5: Strategy Discussion
### Prompt:
> What do you think about the performance of the model and what trading strategies could I build with this model?

### LLM Response:
- The model shows a moderate performance with an accuracy of approximately 60.85%. However, the precision and recall metrics indicate that while the model is good at identifying days when the stock price does not increase (class 0), it struggles significantly to identify days when the stock price does increase (class 1). This suggests a need for further feature engineering or model tuning to improve the prediction performance for this specific classification task.
- Potential Trading Strategies
Given the model's performance, potential trading strategies could be formulated as follows:

#### Conservative Strategy:

- **Focus**: Only trade on days when the model predicts a price increase with high confidence.
- **Action**: Buy stocks when the model predicts an increase (class 1).
- **Risk Management**: Since the model has a high precision for the no-increase class (class 0), it might be more reliable to stay out of the market when the model predicts no increase.

#### Contrarian Strategy:

- **Focus**: Take advantage of the model's high recall for no price increase predictions (class 0).
- **Action**: Sell short or avoid buying stocks on days when the model predicts no increase.
- **Risk Management**: This strategy assumes that the model’s predictions of no price increase are more reliable. It requires a careful assessment of market conditions to avoid false signals.

#### Hybrid Strategy:

- **Combine**: Use both conservative and contrarian strategies based on confidence thresholds.
- **Action**: Set a higher threshold for buying (only buy when the predicted probability of price increase is significantly high) and a lower threshold for selling or staying out of the market.
- **Risk Management**: Adjust thresholds dynamically based on recent model performance and market volatility.


---
## Conclusion

Through this assignment, we have explored how LLMs can assist in complex machine learning tasks without requiring extensive coding knowledge. Key takeaways:

- **Context Management**: LLMs have limited memory, so breaking tasks into smaller steps and reinforcing context is crucial.
- **Feature Engineering with LLMs**: The model can suggest features but needs guidance on financial relevance.
- **Modeling with LLMs**: LLMs can write and refine code, but human judgment is necessary to validate results.
- **Real-world Limitations**: Stock price prediction remains challenging; models should be used critically and not solely relied upon for trading decisions.

Next Steps:
- Experiment with different LLMs and compare outputs.
- Try refining feature engineering techniques.
- Test alternative modeling approaches, such as deep learning or ensemble methods.



In [5]:
# Calculate moving averages
data['MA_5'] = data['Close'].rolling(window=5).mean()
data['MA_10'] = data['Close'].rolling(window=10).mean()
data['MA_50'] = data['Close'].rolling(window=50).mean()

# Calculate daily returns
data['Daily_Return'] = data['Close'].pct_change()

# Drop initial rows with NaN values resulting from rolling calculations
data.dropna(inplace=True)

# Display the first few rows with new features
print(data.head())

Price           Close       High        Low       Open     Volume       MA_5  \
Ticker           AAPL       AAPL       AAPL       AAPL       AAPL              
Date                                                                           
2020-03-13  67.369148  67.841756  61.305270  64.199067  370732000  65.589735   
2020-03-16  58.702309  62.790937  58.166689  58.639292  322423600  64.428341   
2020-03-17  61.283455  62.434667  57.778910  59.986821  324056000  62.853960   
2020-03-18  59.783237  60.590301  57.468687  58.110947  300233600  61.459894   
2020-03-19  59.325176  61.278607  58.799253  59.957739  271857200  61.292665   

Price           MA_10      MA_50 Daily_Return  
Ticker                                         
Date                                           
2020-03-13  68.490310  74.025908     0.119808  
2020-03-16  67.118546  73.747537    -0.128647  
2020-03-17  66.234896  73.534910     0.043970  
2020-03-18  64.875979  73.280818    -0.024480  
2020-03-19  63.709250  

In [6]:
# Create the target variable
data['Price_Increase'] = (data['Close'].shift(-1) > data['Close']).astype(int)

# Drop the last row as the target variable will be NaN
data.dropna(inplace=True)

# Define features (X) and target (y)
features = ['Open', 'High', 'Low', 'Close', 'Volume', 'MA_5', 'MA_10', 'MA_50']
target = 'Price_Increase'

X = data[features]
y = data[target]

print("Features (X) shape:", X.shape)
print("Target (y) shape:", y.shape)

Features (X) shape: (707, 8)
Target (y) shape: (707,)


In [7]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Split data into training and testing sets (70/30 ratio)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

print("Training data shape:", X_train.shape)
print("Testing data shape:", X_test.shape)

Training data shape: (494, 8)
Testing data shape: (213, 8)


In [8]:
# Initialize and train the Random Forest Classifier
model = RandomForestClassifier(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
class_report = classification_report(y_test, y_pred)

print(f"Accuracy: {accuracy:.4f}")
print("\nConfusion Matrix:\n", conf_matrix)
print("\nClassification Report:\n", class_report)

# Report feature importance
feature_importances = pd.Series(model.feature_importances_, index=features).sort_values(ascending=False)
print("\nFeature Importance:\n", feature_importances)

Accuracy: 0.4601

Confusion Matrix:
 [[42 62]
 [53 56]]

Classification Report:
               precision    recall  f1-score   support

           0       0.44      0.40      0.42       104
           1       0.47      0.51      0.49       109

    accuracy                           0.46       213
   macro avg       0.46      0.46      0.46       213
weighted avg       0.46      0.46      0.46       213


Feature Importance:
 High      0.131976
Open      0.127914
MA_5      0.126712
MA_10     0.126592
Volume    0.125274
MA_50     0.123301
Close     0.119699
Low       0.118532
dtype: float64


In [9]:
# Download historical data for AAPL for the prediction period (2023-01-01 to 2024-01-01)
prediction_data = yf.download("AAPL", start="2023-01-01", end="2024-01-01")

# Calculate the same features as used for training
prediction_data['MA_5'] = prediction_data['Close'].rolling(window=5).mean()
prediction_data['MA_10'] = prediction_data['Close'].rolling(window=10).mean()
prediction_data['MA_50'] = prediction_data['Close'].rolling(window=50).mean()

# Create the target variable for evaluation
prediction_data['Price_Increase'] = (prediction_data['Close'].shift(-1) > prediction_data['Close']).astype(int)

# Drop initial and last rows with NaN values
prediction_data.dropna(inplace=True)

# Define features (X_pred) and target (y_pred_true) for the prediction period
X_pred = prediction_data[features]
y_pred_true = prediction_data[target]

print("Prediction features (X_pred) shape:", X_pred.shape)
print("Prediction target (y_pred_true) shape:", y_pred_true.shape)

/tmp/ipython-input-9-1735983965.py:2: FutureWarning: YF.download() has changed argument auto_adjust default to True
  prediction_data = yf.download("AAPL", start="2023-01-01", end="2024-01-01")
[*********************100%***********************]  1 of 1 completed

Prediction features (X_pred) shape: (201, 8)
Prediction target (y_pred_true) shape: (201,)


In [10]:
# Make predictions on the new data
y_pred_new = model.predict(X_pred)

# Evaluate the model on the new data
accuracy_new = accuracy_score(y_pred_true, y_pred_new)
conf_matrix_new = confusion_matrix(y_pred_true, y_pred_new)
class_report_new = classification_report(y_pred_true, y_pred_new)

print(f"Accuracy on new data: {accuracy_new:.4f}")
print("\nConfusion Matrix on new data:\n", conf_matrix_new)
print("\nClassification Report on new data:\n", class_report_new)

Accuracy on new data: 0.4826

Confusion Matrix on new data:
 [[87  5]
 [99 10]]

Classification Report on new data:
               precision    recall  f1-score   support

           0       0.47      0.95      0.63        92
           1       0.67      0.09      0.16       109

    accuracy                           0.48       201
   macro avg       0.57      0.52      0.39       201
weighted avg       0.58      0.48      0.37       201



## Step 5: Strategy Discussion

Based on the performance of the Random Forest model, we can discuss potential trading strategies.

**Model Performance Summary:**

- **Accuracy on Training Data (2020-2022):** 46.01%
- **Accuracy on New Data (2023):** 48.26%

The accuracy on both the training and the new data is close to 50%, which is essentially random chance. The classification report on the new data shows that while the model has a high precision (0.67) when it predicts a price increase (class 1), it has a very low recall (0.09) for this class. This means that when the model predicts an increase, it's often correct, but it misses most of the actual price increases. Conversely, the model has a high recall (0.95) for predicting no price increase (class 0), but a lower precision (0.47). This means it correctly identifies most days with no price increase, but it also incorrectly predicts no increase on many days when the price actually goes up.

**Potential Trading Strategies (with caution):**

Given the model's limitations, any trading strategies should be approached with extreme caution and not solely rely on this model's predictions.

1.  **Conservative Approach (Focus on Precision for Class 1):**
    *   Only consider buying when the model confidently predicts a price increase (class 1).
    *   However, due to the low recall for class 1, this strategy would miss many opportunities for actual price increases.

2.  **Avoiding Downswings (Focus on Recall for Class 0):**
    *   Consider avoiding buying or potentially selling when the model predicts no price increase (class 0), as the model is better at identifying these instances.
    *   This might help in avoiding some potential losses, but the lower precision for class 0 means you might also miss out on some upward movements.

**Critical Thinking:**

The current model's performance suggests it is not a reliable tool for making definitive trading decisions. An accuracy close to 50% indicates that the model is not significantly better than a coin flip at predicting the price trend. The low recall for price increases (class 1) is particularly concerning for a strategy that aims to profit from upward movements.

To potentially build a more useful model, further steps would be needed, such as:
-   More extensive feature engineering, incorporating a wider range of technical indicators, market sentiment, or economic factors.
-   Experimenting with different machine learning models or deep learning architectures.
-   Hyperparameter tuning of the chosen model.
-   Using a longer historical data period for training.

It's crucial to remember that predicting stock prices is inherently challenging due to the multitude of influencing factors and the noisy nature of financial markets.